go/obscuronode/enclave/sql/edgelessdb.go

```diff
@@ -0,0 +1,480 @@
+	if err != nil {
+		return nil, fmt.Errorf("remote attestation of edgeless DB failed - %w", err)
+	}
+	log.Info("retrieved edb PEM: %s", edbPEM)
+
+	// IF DEBUGGING: it can be useful to persist the edb.pem so you can connect to edb from mysql-client on the container
+	//_ = sealAndPersist(edbPEM, dataDir+"edb.pem")
+
+	// now we know we are talking to a secure enclave, we can get the manifest and connect (or initialise if first time)
+	manifest, err := readManifestIfExists()
+	if err != nil && os.IsNotExist(err) {
+		// this is the first time we have connected to this EDB, we will create certificates and a manifest to initialise it
+		log.Info("No manifest found, creating one and initializing edb")
+		manifest, err = createManifestAndInitEDB(edbCfg.Host, edbPEM)
+		if err != nil {
+			return nil, err
+		}
+
+		// Note: it usually takes around 10-15 seconds for edb to initialise and restart
+		log.Info("Waiting 30 seconds for EDB restart after initialization...")
+		time.Sleep(30 * time.Second)
+	} else if err != nil {
+		// this doesn't happen if the manifest file just didn't exist, maybe there was an IO error
+		return nil, fmt.Errorf("failed to read manifest file - %w", err)
+	}
+
+	// we check that this edgeless DB was initialized with the manifest we expected (which is only known to this enclave)
+	log.Info("Validating edb signature against expected manifest...")
+	err = verifyEdgelessDB(edbCfg.Host, manifest, edbPEM)
+	if err != nil {
+		return nil, err
+	}
+
+	// connect to EDB (standard mysql-type connection, using certificate derived from the CA cert in the manifest)
+	log.Info("Setting up SQL connection to edb...")
+	edbSQL, err := connectToEdgelessDB(edbCfg.Host, edbPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to EdgelessDB - %w", err)
+	}
+
+	// wrap it in our eth-compatible key-value store layer
+	return CreateSQLEthDatabase(edbSQL)
+}
+
+// perform the SGX enclave attestation to verify edb running in a legit enclave and with expected edb version etc.
+func performRemoteAttestation(edbHost string) (string, error) {
+	// we need to make sure this dir exists before we start read/writing files in there
+	err := os.MkdirAll(dataDir, 0o644)
+	if err != nil {
+		return "", err
+	}
+	err = prepareEDBAttestationConf(attestationCfgFilepath)
+	if err != nil {
+		return "", fmt.Errorf("failed to prepare latest edb attestation config file - %w", err)
+	}
+	log.Info("Verifying attestation from edgeless DB...")
+	edbHTTPAddr := fmt.Sprintf("%s:%s", edbHost, edbHTTPPort)
+	certs, tcbStatus, err := GetCertificate(edbHTTPAddr, attestationCfgFilepath)
+	if err != nil {
+		// todo should we check the error type with: err == attestation.ErrTCBLevelInvalid?
+		// for now it's maximum strictness (we can revisit this and permit some tcbStatuses if desired)
+		return "", fmt.Errorf("attestation failed, host=%s, tcbStatus=%s, err=%w", edbHTTPAddr, tcbStatus, err)
+	}
+	if len(certs) == 0 {
```

Comment:
Yeah it should be the last cert in the list apparently that is the CA cert, I've always seen it as a single cert but I'll update it to be last cert in the list in case that changes in future edb versions